rfd/0087-access-request-notification-routing.md

```diff
@@ -0,0 +1,321 @@
+access-requests: `targets`.
+
+An example access_request_routing_rule would look like:
+
+```yaml
+kind: access_request_routing_rule
+version: v1
+metadata:
+  name: example
+spec:
+  targets:
+  # simplified syntax
+  - condition: 'resource.spec.roles.contains("prod-rw")'
+    recipients: ["Teleport Alice"]
+    plugin: "pagerduty"
+  # raw expression
+  - expression: >
+      ifelse(
+        resource.spec.roles.contains("prod-rw"),
+        pair(),
+        pair("msteams", set("alice@example.com"))
+      )
+```
+
+Invoked with the role `"prod-rw"`, this would produce the following `access-request`
+
+```yaml
+kind: access_request
+spec:
+  user: bob
+  roles: ["prod-rw"]
+  state: 1
+  # [some fields were omitted for clarity]
+  request_reason: "Prod is down and rollback failed"
+  targets:
+    - plugin: pagerduty
+      recipients: ["Teleport Alice"]
+```
+
+From the plugin point of view, the recipients are additive:
+- they can come from the `role_to_recipient` map in the plugin configuration;
+- they can come from the `targets` on the access-request;
+- they can come from the requests additional recipients; and
+- they can come from annotations (backward compatibility for pagerduty plugin).
+
+### Performance
+
+All `access_request_routing_rules` have to be evaluated against each access request. Even if this only applies
+on access request creation, this can be an intensive operation, especially against the database.
+The auth server should maintain a cache of all target rules to avoid having to request them for each request.
+
+### Security
+
+In case of conflicts (teamA writes a routing rule matching access-requests from teamB), both targets
+will be added to the access request. A misconfigured rule will result in a team being wrongly notified.
+The main issue will be spam/noise but this should not lock out the rightful team to get notified by
+the access request.
+
+Misconfigured rules should not represent security issues unless the plugin does some kind of automatic approval.
+This is the case for the pagerduty plugin which auto-approves if the requestor is on-call for the notified service.
+In such situation, `access_request_routing_rules` should be considered as critical as `roles` themselves as they can lead
+to automatic access escalation. Such risk is already present in the current implementation.
+
+Such risk can be mitigated by not granting the access-request edition rights to the plugin (at the price of a feature loss).
+
+If the two targets are identical it is the plugin's responsibility to deduplicate before sending notifications
+(to avoid sending duplicates and to ensure `PluginData` unity).
+
+### User Experience
+
+#### Complexity/Time to value
+
+In most cases, a simple role_name check is performed and the users wants to set the plugin and the recipients.
+This is provided by the simplified syntax:
+
+```yaml
+targets:
+  - condition: 'resource.spec.roles.contains("prod-rw")'
```

Comment:
this would be more robust if you treated this as `switch` behavior in go, there is no case fall-through, when case matches, it stops

Reply:
If I understand correctly, each `access_request_routing_rule` can provide at most one target, but users can rely on multiple `access_request_routing_rule` to add multiple targets.